node-test/src/test/scala/org/bitcoins/node/networking/peer/MerkleBuffersTest.scala

```diff
@@ -0,0 +1,82 @@
+package org.bitcoins.node.networking.peer
+
+import org.bitcoins.testkit.util.BitcoinSUnitTest
+import org.bitcoins.testkit.Implicits._
+import org.bitcoins.core.protocol.blockchain.MerkleBlock
+import org.bitcoins.testkit.core.gen.BlockchainElementsGenerator
+import org.bitcoins.testkit.core.gen.TransactionGenerators
+import org.scalacheck.Gen
+import org.bitcoins.core.protocol.transaction.Transaction
+import org.bitcoins.core.protocol.blockchain.Block
+import _root_.org.scalatest.compatible.Assertion
+import scala.concurrent.Future
+import scala.util.Success
+import scala.util.Try
+import scala.util.Failure
+
+class MerkleBuffersTest extends BitcoinSUnitTest {
+  behavior of "MerkleBuffers"
+
+  /** Generating blocks and transactions take a little while,
+    * this is to prevent the test from taking a _really_ long
+    * time
+    */
+  implicit override val generatorDrivenConfig: PropertyCheckConfiguration =
+    customGenDrivenConfig(executions = 3)
+
+  it must "match a merkle block with its corresponding transactions" in {
+
+    val txsAndBlockGen: Gen[(Seq[Transaction], Seq[Transaction], Block)] = for {
+      txs <- Gen.nonEmptyListOf(TransactionGenerators.transaction)
+      otherTxs <- Gen.nonEmptyListOf(TransactionGenerators.transaction)
+      block <- BlockchainElementsGenerator.block(txs)
+    } yield (txs, otherTxs, block)
+
+    forAll(txsAndBlockGen) {
+
+      case (txs, otherTxs, block) =>
+        var receivedExpectedTXs: Option[Try[Assertion]] = None
+        var callbackCount: Int = 0
+        val callback: DataMessageHandler.OnMerkleBlockReceived = {
+          (_, merkleTxs) =>
+            receivedExpectedTXs = Some(
+              Try(
+                assert(txs == merkleTxs,
+                       "Received TXs in callback was not the ones we put in")))
+            callbackCount = callbackCount + 1
+        }
+
+        val merkle = MerkleBlock(block, txs.map(_.txId))
+        val _ = MerkleBuffers.putMerkle(merkle)
+
+        txs.map { tx =>
+          val matches = MerkleBuffers.putTx(tx, Seq(callback))
+          assert(
+            matches,
+            s"TX ${tx.txIdBE} did not match any merkle block in MerkleBuffers")
+        }
+
+        otherTxs.map { tx =>
+          val matches = MerkleBuffers.putTx(tx, Seq(callback))
+          assert(
+            !matches,
+            s"Unrelated TX ${tx.txIdBE} did match merkle block in MerkleBuffers")
+
+        }
+
+        assert(callbackCount != 0,
+               "Callback was not called after processing all TXs!")
+
+        assert(callbackCount == 1,
+               s"Callback was called multiple times: $callbackCount")
+
+        receivedExpectedTXs match {
+          case None                     => fail("Callback was never called")
+          case Some(Success(assertion)) => assertion
+          case Some(Failure(exc))       => fail(exc)
+        }
+
+    }
+
```

Comment:
Perhaps add a negative test case here where nothing is matched? What is the expected behavior there?

Reply:
We do test for negative matches by generating a bunch of unrelated transactions, though. Or are you referring to something else?